File: backend/src/models/ServerInvitation.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IServerInvitation extends Document {
  serverId: mongoose.Types.ObjectId;
  senderId: mongoose.Types.ObjectId;
  receiverId: mongoose.Types.ObjectId;
  status: 'pending' | 'accepted' | 'declined';
  expiresAt?: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

const ServerInvitationSchema = new Schema<IServerInvitation>(
  {
    serverId: {
      type: Schema.Types.ObjectId,
      ref: 'Server',
      required: true,
      index: true,
    },
    senderId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    receiverId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
    },
    status: {
      type: String,
      enum: ['pending', 'accepted', 'declined'],
      default: 'pending',
      index: true,
    },
    expiresAt: {
      type: Date,
      default: null,
    },
  },
  {
    timestamps: true,
  }
);

export const ServerInvitation = mongoose.model<IServerInvitation>('ServerInvitation', ServerInvitationSchema);

File: backend/app/models/server_invitation.py
from typing import Optional
from datetime import datetime
from pydantic import Field
from beanie import Document
from beanie import PydanticObjectId as ObjectId
from pymongo import IndexModel, ASCENDING

class ServerInvitation(Document):
    serverId: ObjectId
    senderId: ObjectId
    receiverId: ObjectId
    status: str = "pending" # 'pending', 'accepted', 'declined'
    expiresAt: Optional[datetime] = None
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "serverinvitations"
        indexes = [
            IndexModel([("serverId", ASCENDING)]),
            IndexModel([("receiverId", ASCENDING)]),
            IndexModel([("status", ASCENDING)]),
        ]
